distributed/deploy/spec.py

```diff
@@ -0,0 +1,198 @@
+import asyncio
+
+import toolz
+from tornado import gen
+
+from .cluster import Cluster
+from ..utils import LoopRunner, silence_logging
+from ..scheduler import Scheduler
+
+
+class SpecCluster(Cluster):
+    """ Cluster that requires a full specification of workers
+
+    This attempts to handle much of the logistics of cleanly setting up and
+    tearing down a scheduler and workers, without handling any of the logic
+    around user inputs.  It should form the base of other cluster creation
+    functions.
+
+    Parameters
+    ----------
+    workers: dict
+        A dictionary mapping names to worker classes and their specifications
+        See example below
+    Scheduler: dict, optional
+        A similar mapping for a scheduler
+    asynchronous: bool
+        If this is intended to be used directly within an event loop with
+        async/await
+
+    Examples
+    --------
+    >>> spec = {
+    ...     'my-worker': {"cls": Worker, "options": {"ncores": 1}},
+    ...     'my-nanny': {"cls": Nanny, "options": {"ncores": 2}},
```

Comment:
Sorry - this is probably a dumb question, but why would you want to allocate 2 cores to the nanny process but only 1 to the worker? I thought the nanny didn't do any real work and hence didn't need to be allocated significant resources?

Reply:
Nanny is an object that starts a worker that is watched by a nanny process.  The keywords pass down to the worker.

Reply:
Thanks for the clarification! So the ncores are then shared between the nanny and the worker so it would be ok to only allocate 1 core to a `Nanny` so you could have as many workers as CPUs on the box?

Reply:
The ncores keyword specifies how many threads to add to the thread pool executor of a worker.  This keyword is passed through the nanny to the worker.  This is unrelated to this PR and is just how Nanny works.  You might find it interesting to look through the nanny.py file.